### PR TITLE
Detect parenthesized composition

### PR DESCRIPTION
**Pull request by lue-bird**
Fully implements #209

### CODE REVIEW

File: tests/SimplifyTest.elm [hunk 1]
```diff
@@ -6620,6 +6634,54 @@ a = String.fromList << (String.toList << g << f)
...
+                        [ Review.Test.error
+                            { message = "String.toList, then String.fromList cancels each other out"
+                            , details = [ "You can remove these two functions." ]
+                            , under = "String.fromList"
+                            }
+                            |> Review.Test.whenFixed """module A exposing (..)
+a = (i << h) << (g << f)
+"""
+                        ]
+        , test "should replace (String.fromList >> h >> i) << (f >> g >> String.toList) by (h >> i) << (f >> g)" <|
```
Comment (jfmengels) [inline tests/SimplifyTest.elm:6655]: These are beautiful :heart:

File: src/Simplify.elm [hunk 2]
```diff
@@ -1976,21 +1976,30 @@ expressionVisitorHelp (Node expressionRange expression) config context =
...
+            let
+                innerComposition :
+                    { earlier :
+                        { node : Node Expression, removeRange : Range }
+                    , later :
+                        { node : Node Expression, removeRange : Range }
+                    , isEmbeddedInComposition : Bool
+                    }
+                innerComposition =
+                    getInnerComposition { earlier = compositionSpecific.earlier, later = compositionSpecific.later }
```
Comment (jfmengels) [inline src/Simplify.elm:1993]: This would work as well right?
```suggestion
                    getInnerComposition compositionSpecific
```

Comment (lue-bird) [reply src/Simplify.elm]: very true, fixed.

File: src/Simplify.elm [hunk 3]
```diff
@@ -8433,17 +8395,15 @@ mapAlwaysCompositionChecks :
     -> CompositionIntoCheckInfo
     -> Maybe ErrorInfoAndFix
 mapAlwaysCompositionChecks wrapper checkInfo =
-    case ( checkInfo.earlier.fn, checkInfo.earlier.args ) of
-        ( ( [ "Basics" ], "always" ), [] ) ->
+    case ( ( checkInfo.earlier.fn, checkInfo.earlier.args ), checkInfo.later.args ) of
+        ( ( ( [ "Basics" ], "always" ), [] ), [] ) ->
```
Comment (jfmengels) [inline src/Simplify.elm:8399]: Could use a tuple with 3 elements.
```suggestion
    case ( checkInfo.earlier.fn, checkInfo.earlier.args, checkInfo.later.args ) of
        ( ( [ "Basics" ], "always" ), [], [] ) ->
```

**Review by jfmengels — APPROVED** [review]
LGTM, great work :blush:

File: src/Simplify.elm (hunk 3 above)
Comment (lue-bird) [reply src/Simplify.elm:8399]: For performance? I think I prefer the grouping of tuples :shrug:

Comment (jfmengels) [reply src/Simplify.elm:8399]: I was thinking mostly for readability :thinking: (though yes, it's one less tuple allocation)
Not sure why you find nested tuples to read better, but sure, I don't mind :+1:

**Pull request merged**